content/blog/2023/2024-03-15-getting-started-with-spring-security-and-jwt.md

```diff
@@ -0,0 +1,965 @@
+                version = "1.0.0",
+                license = @License(
+                        name = "Apache 2.0",
+                        url = "http://www.apache.org/licenses/LICENSE-2.0"
+                )),
+        security = {
+                @SecurityRequirement(
+                        name = "bearerAuth"
+                )
+        }
+        )
+@SecurityScheme(
+        name = "bearerAuth",
+        description = "JWT Authorization",
+        scheme = "bearer",
+        type = SecuritySchemeType.HTTP,
+        bearerFormat = "JWT",
+        in = SecuritySchemeIn.HEADER
+)
+public class OpenApiConfig {
+}
+
+````
+Here, security is described using one or more `@SecurityScheme`. The `type` defined here `SecuritySchemeType.HTTP` applies to both basic auth and JWT.
+The other attributes like `scheme` and `bearerFormat` depend on this `type` attribute.
+After defining the security schemes, we can apply them to the whole application or individual operations by adding the 
+`security` section on the root level or operation level.
+In our example, all API operations will use the bearer token authentication scheme.
+For more information, on configuring multiple security schemes and applying a different scheme at the API level, refer to its [documentation.](https://swagger.io/docs/specification/authentication/).
+
+Next, let's add some basic swagger annotations to our controller classes, in order to add descriptions to the API operations.
+````java
+@RestController
+@Tag(name = "Library Controller", description = "Get library books")
+public class BookController {
+}
+
+@RestController
+@Tag(name = "Create Token", description = "Create Token")
+public class TokenController {
+}
+
+````
+
+Also, we will use the below property to override the URL where Springdoc's Swagger-UI loads.
+````yaml
+springdoc:
+  swagger-ui:
+    path: /swagger-ui
+````
+With this configuration, swagger ui will now be available at `http://localhost:8083/swagger-ui/index.html`
+
+Let's try to run the application and load the swagger page at the mentioned URL. When we try to hit the endpoint, we see this:
+{{% image alt="settings" src="images/posts/spring-security-and-jwt/swagger-err.png" %}}
+**This is because all endpoints in the application are automatically secured. We need a way to explicitly exclude the swagger endpoint from being secured.**
+We can do this by adding the `WebSecurityCustomizer` bean and excluding the swagger endpoints in our `SecurityConfiguration` class.
+````java
+@Bean
+    public WebSecurityCustomizer webSecurityCustomizer() {
+        return web -> web.ignoring().antMatchers(
+                ArrayUtils.addAll(buildExemptedRoutes()));
+    }
+
+    private String[] buildExemptedRoutes() {
+        return new String[] {"/swagger-ui/**","/v3/api-docs/**"};
+    }
+````
+Now, when we run the application, the swagger page will load as below:
+{{% image alt="settings" src="images/posts/spring-security-and-jwt/swagger.png" %}}
```

Comment:
You are exposing your bookmarks. Maybe you can remove the bookmark toolbar when taking a screenshot. :-)

Reply:
Done